### ISSUE TITLE
Wrong type with lazy_static leads to rls in VSCode consuming a lot of CPU and getting stuck

### ISSUE THREAD

**Issue by tanyav2** [rust-analyzer/rust-analyzer#6727]
When a `lazy_static` array is initialized with a type signature containing something of the form of a lambda (such as `| |`), rust-analyzer gets stuck indexing (in the examples below, at `indexing: 11/14 my_dir`) and eats up a lot of CPU, the only way to make it stop is to remove the affected lines of code. Understandably this only happens because the code is wrong, but I don't think `rust-analyzer` should hang in this case. 

This happens with and without enabling `rust-analyzer.cargo.loadOutDirsFromCheck` and `rust-analyzer.procMacro.enable` and happens across MacOS Mojave and Ubuntu 20.04. 

Here's some code that works fine:
```
use lazy_static::lazy_static;

lazy_static! {
    static ref FILTERS: [u16; 2] = [
        1, 
        2
    ];
}

fn main() {
    println!("Hello, world!");
}
```
If I change some code, 
```
lazy_static! {
    static ref FILTERS: [|u16| -> bool; 2] = [
        1, 
        2
    ];
}
```
it gets stuck indexing.

**Comment by bjorn3** [comment]
Do you use rls or rust-analyzer? The repo for rls is https://github.com/rust-lang/rls

**Comment by tanyav2** [comment]
Rust analyzer

**Comment by bjorn3** [comment]
I did guess it hangs inside chalk.

**Comment by soulmachine** [comment]
Same here, If I use `assert_eq!` inside `lazy_static!`, my vscode will hung at indexing forever.

**Comment by LevitatingBusinessMan** [comment]
Seems like just any error in these macro's will cause rust-analyzer to hang. So that's cool

**Comment by detrumi** [comment]
> I did guess it hangs inside chalk.

This never gets to chalk solving, it gets stuck before that.

Looking at some debug output, I see a lot of AstIdMapQuery/ParseQuery/ParseMacroExpansionQuery, so this sounds more like a problem with macro expansion.

**Comment by edwin0cheng** [comment]
triage: Fixed, it should handle in #8005 , feel free to reopen it if needed.
